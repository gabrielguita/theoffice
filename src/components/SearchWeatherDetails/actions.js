import {
  SEARCH_WEATHER,
  SEARCH_WEATHER_SUCCESS,
  SEARCH_WEATHER_FAILURE,
} from "./constants";

export const searchWeather = (payload) => {
  return {
    type: SEARCH_WEATHER,
    payload,
  };
};

export const searchWeatherSuccess = (result) => ({
  type: SEARCH_WEATHER_SUCCESS,
  payload: result,
});
export const searchWeatherFailure = (error) => ({
  type: SEARCH_WEATHER_FAILURE,
  payload: error,
});
