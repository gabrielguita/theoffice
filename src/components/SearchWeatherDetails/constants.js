export const SEARCH_WEATHER = "SEARCH_WEATHER";
export const SEARCH_WEATHER_SUCCESS = "SEARCH_WEATHER_SUCCESS";
export const SEARCH_WEATHER_FAILURE = "SEARCH_WEATHER_FAILURE";
