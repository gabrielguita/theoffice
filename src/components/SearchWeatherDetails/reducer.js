import {
  SEARCH_WEATHER,
  SEARCH_WEATHER_SUCCESS,
  SEARCH_WEATHER_FAILURE,
} from "./constants";

export const initialState = {
  weather: {},
  error: null,
  isLoading: false,
};

const weatherReducer = (state = initialState, action) => {
  switch (action.type) {
    case SEARCH_WEATHER:
      return {
        ...state,
        isLoading: true,
      };
    case SEARCH_WEATHER_SUCCESS:
      return {
        ...state,
        isLoading: false,
        weather: action.payload,
      };
    case SEARCH_WEATHER_FAILURE:
      return {
        ...state,
        isLoading: true,
        error: action.payload,
      };

    default:
      return state;
  }
};

export default weatherReducer;
