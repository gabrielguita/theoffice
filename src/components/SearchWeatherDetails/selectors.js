import { createSelector } from "reselect";
import get from "lodash/get";

const selectShow = (state) => state;

const makeSelectShow = () => createSelector(selectShow, (substate) => substate);

const makeSelectGetWeather = () =>
  createSelector(selectShow, (substate) => ({
    isLoading: get(substate, "getWeather.isLoading", null),
    hasError: get(substate, "getWeather.weather.message", null),
    headline: get(substate, "getWeather.weather.Headline.text", null),
    category: get(substate, "getWeather.weather.Headline.Category", null),
    dailyForecasts: get(substate, "getWeather.weather.DailyForecasts", null),
  }));

export { makeSelectShow, makeSelectGetWeather };
