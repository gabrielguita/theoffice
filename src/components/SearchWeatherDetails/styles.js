import styled from "styled-components";
import media from "styled-media-query";

export const Wrapper = styled.div`
  color: #000;
  font-size: 12px;
  ${media.lessThan("375px")`
    margin: 15px 0 0 0;
  `}
`;

export const ShowWeather = styled.div`
  color: #000;
  cursor: pointer;
  font-weight: bold;
  max-width: 220px;
  font-size: 14px;
  text-decoration: underline;
  &:hover {
    opacity: 0.7;
  }
`;
export const Forecast = styled.div`
  color: #000;
  font-size: 12px;
`;

export const Title = styled.h2`
  color: blue;
`;

export const Category = styled.h4`
  color: #000;
`;

export const DailyForecasts = styled.ul`
  color: #000;
  font-size: 12px;
  li {
    list-style: none;
  }
  ${media.lessThan("375px")`
    margin: 15px 0 0 0;
  `}
`;
