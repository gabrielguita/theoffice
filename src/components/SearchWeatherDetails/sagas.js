import { put, all, call, takeLatest } from "redux-saga/effects";
import { searchWeatherSuccess, searchWeatherFailure } from "./actions";
import { SEARCH_WEATHER } from "./constants";

import fetchWeatherServices from "../../services/weatherServices";

export function* searchWeather({ payload }) {
  try {
    const apiGetResponse = yield call(fetchWeatherServices, payload.code);
    yield put(searchWeatherSuccess(apiGetResponse));
  } catch (error) {
    yield put(searchWeatherFailure(error));
  }
}

export default function* rootSagas() {
  yield all([takeLatest(SEARCH_WEATHER, searchWeather)]);
}
