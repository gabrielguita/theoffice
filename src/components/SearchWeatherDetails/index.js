import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import moment from "moment";
import {
  Wrapper,
  Forecast,
  Title,
  Category,
  DailyForecasts,
  ShowWeather,
} from "./styles";
import { makeSelectGetWeather } from "./selectors";
import { createStructuredSelector } from "reselect";
import { searchWeather as searchWeatherAction } from "./actions";
import Spinner from "../../components/Spinner/";
import ErrorMessages from "../../components/ErrorMessages";

const getCity = (data) => {
  switch (data) {
    case "BUD":
      return { city: "Budapest", code: 187423 };
    case "AMS":
      return { city: "Amsterdam", code: 249758 };
    case "MAD":
      return { city: "Madrid", code: 308526 };

    default:
      return false;
  }
};

const toCelsius = (f) => Math.trunc((5 / 9) * (f - 32));
const formatDate = (date) => moment(date).format("DD/MM/YYYY");
const forecastIcon = (i) => {
  return i.toString().length === 1 ? `${"0" + i}` : i;
};

const SearchWeatherDetails = (props) => {
  const { searchWeather, data, getWeather } = props;
  const {
    headline,
    category,
    dailyForecasts,
    isLoading,
    hasError,
  } = getWeather;
  const handleOnCLick = () => {
    searchWeather(getCity(data));
  };
  return (
    <Wrapper>
      <ShowWeather onClick={handleOnCLick}>
        Check the weather in: {getCity(data).city}
      </ShowWeather>
      {hasError && <ErrorMessages msg={hasError} />}
      <Forecast>
        <Title>{headline} </Title>
        <Category>{category}</Category>
        <DailyForecasts>
          {dailyForecasts &&
            dailyForecasts.map((forecast, i) => (
              <li key={i}>
                <p>
                  <strong>Day: </strong> {forecast.Day.PrecipitationType} /{" "}
                  {forecast.Day.PrecipitationIntensity}
                  <img
                    alt="day"
                    src={`https://developer.accuweather.com/sites/default/files/${forecastIcon(
                      forecast.Day.Icon,
                    )}-s.png`}
                  />
                </p>
                <p>
                  <strong>Night: </strong> {forecast.Night.PrecipitationType} /{" "}
                  {forecast.Night.PrecipitationIntensity}
                  <img
                    alt="night"
                    src={`https://developer.accuweather.com/sites/default/files/${forecastIcon(
                      forecast.Night.Icon,
                    )}-s.png`}
                  />
                </p>
                <p>
                  <i>{formatDate(forecast.Date)}</i>
                </p>
                <p>
                  Temperature minimum:{" "}
                  {toCelsius(forecast.Temperature.Minimum.Value)} °C
                </p>
                <p>
                  Temperature maximum:{" "}
                  {toCelsius(forecast.Temperature.Maximum.Value)} °C
                </p>
              </li>
            ))}
        </DailyForecasts>
      </Forecast>
      {isLoading && <Spinner />}
    </Wrapper>
  );
};

SearchWeatherDetails.propTypes = {
  show: PropTypes.object,
  searchWeatherAction: PropTypes.func,
  searchWeather: PropTypes.func,
  isLoading: PropTypes.bool,
  getWeather: PropTypes.object,
};

const mapStateToProps = createStructuredSelector({
  getWeather: makeSelectGetWeather(),
});

export function mapDispatchToProps(dispatch) {
  return {
    searchWeather: (data) => dispatch(searchWeatherAction(data)),
  };
}

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(SearchWeatherDetails);
