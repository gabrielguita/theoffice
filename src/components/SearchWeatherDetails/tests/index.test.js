import React from "react";
import SearchWeatherDetails from "../index";

describe("<SearchWeatherDetails />", () => {
  it("should render correctly", () => {
    expect(<SearchWeatherDetails />).toMatchSnapshot();
  });
});
