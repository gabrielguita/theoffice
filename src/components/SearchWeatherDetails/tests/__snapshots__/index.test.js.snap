// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchWeatherDetails /> should render correctly 1`] = `<UNDEFINED />`;
